### PR TITLE
force quick downscale MIP_F previews and refactor some demosaic choices into a single function

### PR DESCRIPTION
**Pull request by dtorop**
This addresses [bug #11340](https://redmine.darktable.org/issues/11340). When demosaicing MIP_Fs, this change makes previews always be quickly downscaled rather than going through a full demosaic. It also coalesces the logic behind determining whether how to perform the demosaic (whether to perform a full scale demosaic, only a linear interpolation, and a full Markesteijn demosaic) into a single function `demosaic_op_flags()` rather than repeating the same logic in several places in demosaic.c.

The first change appears to eliminate negative values from demosaic for both [Bayer](http://www.rawsamples.ch/raws/canon/RAW_CANON_EOS_5DS.CR2) and [X-Trans](https://dl.dropboxusercontent.com/u/2171814/_DSF1080.RAF) sample images.

It's not entirely clear why full demosaic algorithms fail for these images. A working theory is that when they are transformed in _init_f() from 1-channel full-size to 1-channel MIP_F-size, the noise in dark areas effectively increases frequency, and this leads a full demosaic to interpolate negative values.

NOTE: I updated PR title and description on 1/5/17 to reflect that it no longer addresses bug #11167 nor forces lower-resolution output from demosaic iop.

### CODE REVIEW

File: src/iop/demosaic.c [hunk 1]
```diff
@@ -1502,6 +1502,15 @@ void modify_roi_in(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t *
   // this op is disabled for preview pipe/filters == 0
 
   *roi_in = *roi_out;
+
+  if ((piece->pipe->type == DT_DEV_PIXELPIPE_PREVIEW) &&
+      self->dev->image_storage.buf_dsc.filters)
+  {
+    // This is a mosaiced MIP_F, and was given a larger buffer as
+    // 1-channel. Shrink to proper size now that it is 4-channel.
+    roi_in->scale /= 2.0f;
```
Comment (dtorop) [inline src/iop/demosaic.c]: I realize this bit is utterly wrong, and doesn't in fact halve the image size as hoped for. Still looking for the right way to do this.

**Comment by LebedevRI** [comment]
> but doesn't eliminate edge artifacts.

Briefly discussed in IRC, not for release.
The final version can always be cherry-picked for next point release.

@dtorop thank you for looking into this!

**Comment by dtorop** [comment]
@LebedevRI: sounds wise. Certainly not worth risking destabilizing the chain of RCs leading to a major version.

I'll work up the sub-sampling for X-Trans, am curious if that helps with the edge artifacts. There are also further improvements to make along the lines of https://github.com/dtorop/darktable/tree/xtrans-better-downscale, well worth my working this up for all the X-Trans paths. I also found a way to slightly clean up the port of Markesteijn demosaic to dt, which should end up in a different PR, but again, that could be something for a point release.

Things have been busy here, but look forward to spending time on this.

**Comment by upegelow** [comment]
@dtorop: I have concerns with the downscaling early in the pixelpipe. You know that the preview is not only used for display purposes but some modules like perspecitve correction need to take its nput when they want to see the full image rather than the potentially zoomed-in center view. Currently the preview is typically like 1350 x 900. If we go down to 720 x 450 early in the pipe a lot of resolution is lost. For perspective correction this means that the accuracy of the automatic fit will drop clearly.

**Comment by dtorop** [comment]
@upegelow It's a good point... I realize that I didn't update the description of the PR (or its title) to the current contents, which I just did. The PR no longer changes the size of the preview (that code wasn't working anyhow). It limits itself to forcing a quick downscale for previews (which should eliminate negative values in dark areas introduced by interpolation in a full demosaic) and refactoring into a single function some repeated logic about what sort of demosaic to perform. Apologies for the mis-described PR.

In dt 2.0, I think the previews were 720x450 right from the go (`_init_f()`) through the whole pipeline. In dt 2.2, they are (I believe) approx 1350x900 through the whole pipeline. Initially I thought it was good to return to the dt 2. behavior via shrinking them in demosaic. But it sounds like we need them 1350x900-ish through the whole pipeline...

**Pull request merged**